lib/trackler/problem.rb

```diff
@@ -7,81 +7,108 @@ class Problem
     end
 
     def exists?
-      !!md && !!yaml
+      description_exists? && yaml_exists?
     end
 
     def name
-      @name ||= slug.split('-').map(&:capitalize).join(' ')
+      @name ||= slug_to_name
     end
 
     def description
-      @description ||= File.read(filepath(md))
+      @description ||= load_description
     end
 
     def source_markdown
-      body = source.empty? ? "" : "%s" % source
-      url = source_url.empty? ? "" : "[%s](%s)" % [source_url, source_url]
-      if url.empty? && body.empty?
-        ""
-      else
-        "## Source\n\n" + [body, url].reject(&:empty?).join(" ")
-      end
+      text = [source, markdown_link(source_url)].reject(&:empty?).join(" ")
+      text.empty? ? text : "## Source\n\n#{text}"
     end
 
     def md_url
-      repo_url(md)
+      repo_url(md_path)
     end
 
     def json_url
-      repo_url(json) if !!json
+      repo_url(json_path) if json_path
     end
 
     def yaml_url
-      repo_url(yaml)
+      repo_url(yaml_path)
     end
 
-    %w(blurb source source_url).each do |name|
-      define_method name do
-        metadata[name].to_s.strip
-      end
+    def blurb
+      metadata['blurb'].to_s.strip
+    end
+
+    def source
+      metadata['source'].to_s.strip
+    end
+
+    def source_url
+      metadata['source_url'].to_s.strip
     end
 
     private
 
-    def json
+    def description_exists?
+      !md_path.nil?
+    end
+
+    def yaml_exists?
+      !yaml_path.nil?
+    end
+
+    def slug_to_name
+      slug.split('-').map(&:capitalize).join(' ')
+    end
+
+    def json_path
       [
         "exercises/%s/canonical-data.json" % slug,
-        "%s.json" % slug,
-      ].find { |path| File.exist?(filepath(path)) }
+      ].find { |filename| common_metadata_file_exists?(filename) }
     end
 
-    def yaml
+    def yaml_path
       [
         "exercises/%s/metadata.yml" % slug,
-        "%s.yml" % slug,
-      ].find { |path| File.exist?(filepath(path)) }
+      ].find { |filename| common_metadata_file_exists?(filename) }
     end
 
-    def md
+    def md_path
       [
         "exercises/%s/description.md" % slug,
-        "%s.md" % slug,
-      ].find { |path| File.exist?(filepath(path)) }
+      ].find { |filename| common_metadata_file_exists?(filename) }
     end
 
     def repo_url(path)
-      "https://github.com/exercism/x-common/blob/master/%s" % path
+      "#{@xcommon_repository_url}/#{path}"
     end
 
-    def filepath(f)
-      File.join(root, "common", f)
+    def metadata
+      @metadata ||= load_yaml
     end
 
-    def metadata
-      @metadata ||= YAML.load(File.read(filepath(yaml)))
+    def common_metadata_path(path)
+      File.join(root, "common", path)
+    end
+
+    def common_metadata_file_exists?(filename)
+      File.exist?(common_metadata_path(filename))
+    end
+
+    def load_description
+      File.read(common_metadata_path(md_path))
+    end
+
+    def load_yaml
+      YAML.load(File.read(common_metadata_path(yaml_path)))
```

Comment:
Same here. I don't think it's worth having a method when the name just echoes the implementation.